### PR TITLE
Add support for xsalsa20_poly1305_lite

### PR DESCRIPTION
**Pull request by swarley**
### Summary

Adds support for `xsalsa20_poly1305_lite` and changes the protocol selection method to support ranked choice.

Also adds `lite_nonce` to `VoiceClient`, used for the incremental nonce.

### Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by imayhaveborkedit** [comment]
Turns out I already did this back when I did the vwsV4 pr and just never pr'd it here for some reason.  

https://github.com/imayhaveborkedit/discord.py/commit/5f15734b5528a1a6d5c54b04889ffb6ddc1b1886

File: discord/voice_client.py [hunk 1]
```diff
@@ -331,6 +333,20 @@ def _encrypt_xsalsa20_poly1305_suffix(self, header, data):
 
         return header + box.encrypt(bytes(data), nonce).ciphertext + nonce
 
+    def _encrypt_xsalsa20_poly1305_lite(self, header, data):
+        box = nacl.secret.SecretBox(bytes(self.secret_key))
+        nonce = bytearray(24)
+
+        # Upper limit of unsigned 4 byte int
+        if self.lite_nonce >= 0xff_ff_ff_ff:
```
Comment (Rapptz) [inline discord/voice_client.py]: This is 3.6 syntax.

File: discord/gateway.py [hunk 2]
```diff
@@ -727,7 +727,7 @@ def __init__(self, *args, **kwargs):
         log.debug('detected ip: %s port: %s', state.ip, state.port)
 
         # there *should* always be at least one supported mode (xsalsa20_poly1305)
-        modes = [mode for mode in data['modes'] if mode in self._connection.supported_modes]
+        modes = list(set(self._connection.supported_modes).intersection(data['modes']))
```
Comment (Rapptz) [inline discord/gateway.py]: I don't see why this change was necessary.

Comment (swarley) [reply discord/gateway.py]: I'm in the process of reverting it. In testing `xsalsa20_poly1305` was still being chosen over preferable modes. However more closely examining results from this change shows this also does not fix that consistently.

File: discord/voice_client.py [hunk 3]
```diff
@@ -111,9 +111,11 @@ def __init__(self, state, timeout, channel):
         self._runner = None
         self._player = None
         self.encoder = None
+        self.lite_nonce = 0
```
Comment (imayhaveborkedit) [inline discord/voice_client.py]: The nonce is an impl detail and doesn't need to be public.

File: discord/voice_client.py [hunk 4]
```diff
@@ -331,6 +333,20 @@ def _encrypt_xsalsa20_poly1305_suffix(self, header, data):
...
+    def _encrypt_xsalsa20_poly1305_lite(self, header, data):
+        box = nacl.secret.SecretBox(bytes(self.secret_key))
+        nonce = bytearray(24)
+
+        # Upper limit of unsigned 4 byte int
+        if self.lite_nonce >= 0xffffffff:
+            self.lite_nonce = 0
+
+        nonce[:4] = struct.pack('>I', self.lite_nonce)
+        self.lite_nonce += 1
```
Comment (imayhaveborkedit) [inline discord/voice_client.py]: This is exactly what the `checked_add` function is for.

Comment (swarley) [reply discord/voice_client.py]: I don't see how adding another function for this conditional is beneficial

Comment (imayhaveborkedit) [reply discord/voice_client.py]: The library already does this here:
https://github.com/Rapptz/discord.py/blob/c5f01eb66efb82fd1d2cf32b3658113322b37c01/discord/voice_client.py#L456
